tests/app/main/test_services.py

```diff
@@ -1030,6 +1058,67 @@ def test_questions_for_this_service_section_can_be_changed(self, data_api_client
             "service_updated",
         )
 
+    def test_file_upload(self, data_api_client, s3):
+        data_api_client.get_service.return_value = self.base_service
+        with freeze_time('2017-11-12 13:14:15'):
+            res = self.client.post(
+                '/suppliers/frameworks/g-cloud-9/services/321/edit/documents',
+                data={
+                    'serviceDefinitionDocumentURL': (StringIO(b'doc'), 'document.pdf'),
+                }
+            )
+
+        assert res.status_code == 302
+        data_api_client.update_service.assert_called_once_with(
+            '321',
+            {
+                'serviceDefinitionDocumentURL':
+                'http://asset-host/g-cloud-9/documents/1234/321-service-definition-document-2017-11-12-1314.pdf'
+            },
+            'email@email.com',
+        )
+
+        s3.return_value.save.assert_called_once_with(
+            'g-cloud-9/documents/1234/321-service-definition-document-2017-11-12-1314.pdf',
+            mock.ANY, acl='public-read'
+        )
+
+    def test_S3_should_not_be_called_if_there_are_no_files(self, data_api_client, s3):
+        data_api_client.get_service.return_value = self.base_service
+        res = self.client.post(
+            '/suppliers/frameworks/g-cloud-9/services/321/edit/service-features-and-benefits',
+            data={
+                'serviceFeatures': ["one", "two"],
+                'serviceBenefits': ["three", "four"],
+            })
+
+        assert res.status_code == 302
+        assert s3.return_value.save.called is False
+
+    def test_file_upload_filters_empty_and_unknown_files(self, data_api_client, s3):
+        data_api_client.get_service.return_value = self.base_service
+        res = self.client.post(
+            '/suppliers/frameworks/g-cloud-9/services/321/edit/documents',
+            data={
+                'serviceDefinitionDocumentURL': (StringIO(b''), 'document.pdf'),
+                'unknownDocumentURL': (StringIO(b'doc'), 'document.pdf'),
+            })
+
+        assert res.status_code == 302
+        assert s3.return_value.save.called is False
+        data_api_client.update_service.assert_called_once_with('321', {}, 'email@email.com')
+
+    def test_upload_question_can_not_be_set_by_form_data(self, data_api_client, s3):
+        data_api_client.get_service.return_value = self.base_service
+        res = self.client.post(
+            '/suppliers/frameworks/g-cloud-9/services/321/edit/documents',
+            data={
+                'serviceDefinitionDocumentURL': 'http://example.com/document.pdf',
+            })
+
+        assert res.status_code == 302
```

Comment:
Don't these three cases (no files, empty/unknown files, url in form data) count as 'errors'? Why do they return 302 instead of 400?

Reply:
That's a good question that I don't necessarily have a good answer to right now.  The trying-to-pass-in-in-as-a-key thing is a weird one - because it's a valid key with a valid value then I'm not sure how we'd make it provoke a validation error.  The ignoring of empty/unknown files is existing behaviour in service submissions that someone had a reason for at some point... it is a bit weird though, people might want to be alerted if they try to upload an empty file. Maybe I'll check the history to see if I can find some rationale...

Reply:
If you can live with it then I'm very tempted to stick with the existing ("legacy") behaviour and leave it as it is for now.  But if you think I'm just being lazy then tell me to fix it.

Reply:
Happy to leave it as long as it's documented (my rule of thumb is that if it's not obvious from the code what's happening or an unusual decision has been made, stick an explanatory comment in). This should save us having this same conversation again the next time we touch this code 😸

Reply:
I added a comment.  Think this should be good to go now?